service-registry/src/main/java/io/servicecomb/serviceregistry/consumer/MicroserviceManager.java

```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright 2017 Huawei Technologies Co., Ltd
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.servicecomb.serviceregistry.consumer;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class MicroserviceManager {
+  private AppManager appManager;
+
+  private String appId;
+
+  // key: microserviceName
+  private Map<String, MicroserviceVersions> versionsByName = new ConcurrentHashMap<>();
+
+  public MicroserviceManager(AppManager appManager, String appId) {
+    this.appManager = appManager;
+    this.appId = appId;
+  }
+
+  public MicroserviceVersionRule getOrCreateMicroserviceVersionRule(String microserviceName,
+      String versionRule) {
+    MicroserviceVersions microserviceVersions =
+        versionsByName.computeIfAbsent(microserviceName, name -> {
```

Comment:
same issue there, I suggest to use name variable in the construction method of MicroserviceVersion.

Reply:
same to above